[namedtensor] fix autograd support for torch.mean(tensor, dimname)

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29199 [namedtensor] fix autograd support for torch.mean(tensor, dimname)**

Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`

Differential Revision: [D18334617](https://our.internmc.facebook.com/intern/diff/D18334617)

zou3519 committed:
[namedtensor] fix autograd support for torch.mean(tensor, dimname)

Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`

zou3519 committed:
Update on "[namedtensor] fix autograd support for torch.mean(tensor, dimname)"


Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`

zou3519 committed:
Update on "[namedtensor] fix autograd support for torch.mean(tensor, dimname)"


Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`

nairbv commented on test/test_namedtensor.py:
```diff
@@ -1054,6 +1062,11 @@ def test_simple_reduce(op, device):
+        def test_autograd_supports_dimname_overload(op, device):
+            t = torch.empty(2, 3, 5, names=('N', 'C', 'L'), device=device, requires_grad=True)
+            sum_all_outputs(op(t, 'C')).backward()
+            self.assertTrue(t.grad is not None)
```
can we do anything to validate the value of the gradient? would it be helpful to validate the grad against the equivalent op on an unnamed tensor?

nairbv replied:
nit: btw I think there's a self.assertIsNotNone()

zou3519 replied:
> can we do anything to validate the value of the gradient?

We can, but see the next point

> would it be helpful to validate the grad against the equivalent op on an unnamed tensor?

Because autograd ignores all names on tensors, we don't need to validate the grad as long as we test that "autograd ignores all names on tensors"

> nit: btw I think there's a self.assertIsNotNone()

Sure I'll use that

nairbv replied:
>autograd ignores all names on tensors 

will that ever not be the case for any kinds of gradients?

I think you had also mentioned an issue with the correctness of the mean gradient that we were also fixing, though maybe that's (going to be) a separate PR?

zou3519 replied:
> > autograd ignores all names on tensors
> 
> will that ever not be the case for any kinds of gradients?

In the future we may design and implement name-aware autograd. This is not going to happen soon. When that happens, we can revisit the testing plan for this.

> I think you had also mentioned an issue with the correctness of the mean gradient that we were also fixing, though maybe that's (going to be) a separate PR?

yeah, that's orthogonal to this PR. It is this one: https://github.com/pytorch/pytorch/issues/29137#issuecomment-549938081

zou3519 replied:
This PR fixes https://github.com/pytorch/pytorch/issues/28286

nairbv replied:
thanks for the clarification

zou3519 committed:
Update on "[namedtensor] fix autograd support for torch.mean(tensor, dimname)"


Previously, we called `native::mean_cpu_gpu` inside `mean(Tensor, Dimname)`;
`native::mean_cpu_gpu` is not supported by autograd. This PR replaces
`native::mean_cpu_gpu` with `at::mean(Tensor, int)` so that the dimname
overload can piggyback off of autograd support for `at::mean(Tensor,
int)`.

Also added tests (those didn't exist before) for autograd support for
named tensor reduction functions.

Test Plan:
- `python test/test_namedtensor.py -v`
